pkg/controller/utils/utils.go

```diff
@@ -161,3 +167,116 @@ func GetDeployment(c client.Client, ns string) (*appsv1.Deployment, error) {
 	}
 	return &d, nil
 }
+
+// CreateOrUpdateSecretIfNotExists creates a secret in case it does not exist or updates it if there are changes
+func CreateOrUpdateSecretIfNotExists(c client.Client, r client.Reader, secretName string, targetNS string, data map[string][]byte, secretType corev1.SecretType, log logr.Logger) error {
+	var cfg corev1.Secret
+	err := r.Get(context.TODO(), client.ObjectKey{Name: secretName, Namespace: targetNS}, &cfg)
+	if errors.IsNotFound(err) {
+		log.Info("Creating OneAgent config secret")
+		if err := c.Create(context.TODO(), &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      secretName,
+				Namespace: targetNS,
+			},
+			Type: secretType,
+			Data: data,
+		}); err != nil {
+			return fmt.Errorf("failed to create secret %s: %w", secretName, err)
+		}
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to query for secret %s: %w", secretName, err)
+	}
+
+	if !reflect.DeepEqual(data, cfg.Data) {
+		log.Info(fmt.Sprintf("Updating secret %s", secretName))
+		cfg.Data = data
+		if err := c.Update(context.TODO(), &cfg); err != nil {
+			return fmt.Errorf("failed to update secret %s: %w", secretName, err)
+		}
+	}
+
+	return nil
+}
+
+// GeneratePullSecretData generates the secret data for the PullSecret
+func GeneratePullSecretData(c client.Client, apm dynatracev1alpha1.OneAgentAPM, tkns corev1.Secret) (map[string][]byte, error) {
+	dtc, err := BuildDynatraceClient(c, &apm)
+	if err != nil {
+		return nil, err
+	}
+
+	ci, err := dtc.GetConnectionInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	registry, err := GetImageRegistryFromAPIURL(apm.Spec.APIURL)
+	if err != nil {
+		return nil, err
+	}
+
+	auth := fmt.Sprintf("%s:%s", ci.TenantUUID, string(tkns.Data[DynatracePaasToken]))
+	auth = b64.StdEncoding.EncodeToString([]byte(auth))
+	dockercfg := fmt.Sprintf("{\"auths\":{\"%s\":{\"username\":\"%s\",\"password\":\"%s\",\"auth\":\"%s\"}}}", registry, ci.TenantUUID, string(tkns.Data[DynatracePaasToken]), auth)
```

Comment:
I'd suggest using `json.Marshal()` here to avoid dealing with unescaped characters.